File: platforms/apple/Sources/BotaAppSDK/ClientPresence.swift
import Foundation

/// Passive SDK metadata. The host may explicitly attach it to an authenticated heartbeat.
public struct SDKClientContext: Equatable, Sendable {
    public let schemaVersion: Int
    public let sessionID: String
    public let sequence: Int64
    public let platform: String
    public let sdkPackage: String
    public let sdkVersion: String
}

public struct ClientPresence: Sendable {
    let devices: DeviceManager

    /// Returns nil without a current verified connection. Does not read GATT or send HTTP.
    public func nextReport(deviceID: String) async throws -> SDKClientContext? {
        await devices.nextClientReport(deviceID: deviceID)
    }
}

/// Confined to DeviceManager's actor; callbacks must name the session they belong to.
struct ConnectionClientPresence {
    private(set) var sessionID: String?
    private(set) var transportID: String?
    private var deviceID: String?
    private var sequence: Int64 = 0
    private var destroyed = false

    mutating func connected(deviceID: String, sessionID: String = UUID().uuidString.lowercased(), transportID: String) {
        guard !destroyed else { return }
        self.deviceID = deviceID
        self.sessionID = sessionID
        self.transportID = transportID
        sequence = 0
    }

    mutating func disconnected(sessionID: String?) {
        guard self.sessionID == sessionID else { return }
        self.sessionID = nil
        transportID = nil
        deviceID = nil
    }

    mutating func destroy() {
        destroyed = true
        disconnected(sessionID: sessionID)
    }

    mutating func nextReport(deviceID: String) -> SDKClientContext? {
        guard !destroyed, self.deviceID == deviceID, let sessionID,
              sequence < 9_007_199_254_740_991 else { return nil }
        sequence += 1
        #if os(iOS)
        let platform = "ios"
        #else
        let platform = "macos"
        #endif
        return SDKClientContext(schemaVersion: 1, sessionID: sessionID, sequence: sequence,
                                platform: platform, sdkPackage: SdkIdentity.package, sdkVersion: SdkIdentity.version)
    }
}

File: frameworks/web/src/clientPresence.ts
import { SDK_PACKAGE, SDK_VERSION } from './sdkIdentity.ts'

/** Self-reported diagnostics, not identity attestation or command authority. */
export interface SdkClientContext {
  schema_version: 1
  session_id: string
  sequence: number
  platform: 'web'
  sdk_package: string
  sdk_version: string
}

export interface ClientPresence {
  /** Local-only. deviceId is the SDK handle, not a backend dev_* identifier. */
  nextReport(deviceId: string): Promise<SdkClientContext | null>
}

/** Internal passive owner. Only verified connection lifecycle calls connected. */
export class ConnectionClientPresence implements ClientPresence {
  private session: { deviceId: string; id: string; sequence: number } | null = null
  private destroyed = false
  private readonly uuid: () => string

  constructor(uuid: () => string = () => globalThis.crypto.randomUUID()) {
    this.uuid = uuid
  }

  connected(deviceId: string): () => void {
    if (this.destroyed) return () => {}
    const session = { deviceId, id: this.uuid(), sequence: 0 }
    this.session = session
    return () => { if (this.session === session) this.session = null }
  }

  async nextReport(deviceId: string): Promise<SdkClientContext | null> {
    const session = this.session
    if (!session || session.deviceId !== deviceId || this.destroyed) return null
    return {
      schema_version: 1, session_id: session.id, sequence: ++session.sequence,
      platform: 'web', sdk_package: SDK_PACKAGE, sdk_version: SDK_VERSION,
    }
  }

  destroy(): void {
    this.destroyed = true
    this.session = null
  }
}
